### ISSUE TITLE
torch.compile and eager expect different shapes of an input to index_copy_

### ISSUE THREAD

**Issue by zielin32** [pytorch/pytorch#115931]
### 🐛 Describe the bug

I see an error when I try to perform an index_copy_ operation with torch.compile. Code:
```python
import torch
device = torch.device('cpu')

# replace [1, 10] with [10] to make it work in torch.compile and fail in eager
x = torch.zeros([1, 10], device=device)
xs = torch.ones([2, 10], device=device)

def index_copy(xs, x):
    xs.index_copy_(0, torch.tensor(0).to(device), x)

# comment out to run in eager mode
index_copy = torch.compile(index_copy)

print("Before index copy")
print(f"x: {x}")
print(f"xs: {xs}")

index_copy(xs, x)

print()
print("After index copy")
print(f"xs: {xs}")
```

When I run the code, I get the following output:
```
Before index copy
x: tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
xs: tensor([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])
---------------------------------------------------------------------------
BackendCompilerFailed                     Traceback (most recent call last)
[<ipython-input-12-461a3130f2da>](https://localhost:8080/#) in <cell line: 18>()
     16 print(f"xs: {xs}")
     17 
---> 18 index_copy(xs, x)
     19 
     20 print()

53 frames
[/usr/local/lib/python3.10/dist-packages/torch/_inductor/ir.py](https://localhost:8080/#) in _normalize_size(x, new_size)
   1504         old_size = x.get_size()
   1505         old_size = [None] * (len(new_size) - len(old_size)) + list(old_size)
-> 1506         assert len(new_size) == len(old_size)
   1507         for i in range(len(new_size)):
   1508             if new_size[i] == -1:

BackendCompilerFailed: backend='inductor' raised:
LoweringException: AssertionError: 
  target: aten.index_put.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg0_1', layout=FixedLayout('cpu', torch.float32, size=[2, 10], stride=[10, 1]))
  ))
  args[1]: [TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.int64,
      def inner_fn(index):
          tmp0 = ops.constant(0, torch.int64)
          return tmp0
      ,
      ranges=[],
      origin_node=full_default,
      origins={full_default}
    )
  ))]
  args[2]: TensorBox(StorageBox(
    InputBuffer(name='arg1_1', layout=FixedLayout('cpu', torch.float32, size=[1, 10], stride=[10, 1]))
  ))

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```
Same code works fine, when I comment out the line with _torch.compile_:
```
Before index copy
x: tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
xs: tensor([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

After index copy
xs: tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])
```
Interestingly, when I change the shape of _x_ from [1. 10] to [10], the code works fine with _torch.compile_ (same output as above), and fails in eager mode:
```
Before index copy
x: tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])
xs: tensor([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
[<ipython-input-18-b9c58744757d>](https://localhost:8080/#) in <cell line: 18>()
     16 print(f"xs: {xs}")
     17 
---> 18 index_copy(xs, x)
     19 
     20 print()

[<ipython-input-18-b9c58744757d>](https://localhost:8080/#) in index_copy(xs, x)
      7 
      8 def index_copy(xs, x):
----> 9     xs.index_copy_(0, torch.tensor(0).to(device), x)
     10 
     11 # comment out to run in eager mode

IndexError: index_copy_(): When source and destination are not scalars, their dimensionality must match. Source dimensionality (1), destination dimensionality (2)
```

It looks like eager and torch.compile aren't entirely compatible and some code changes are necessary in order to switch between the two. In this case, it seems that index_copy_ in eager mode expects the number of dims of both tensors to match, while torch.compile requires a squeeze on the tensor being copied.

### Versions

```
PyTorch version: 2.1.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.9
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.1.58+-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             2
On-line CPU(s) list:                0,1
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7B12
CPU family:                         23
Model:                              49
Thread(s) per core:                 2
Core(s) per socket:                 1
Socket(s):                          1
Stepping:                           0
BogoMIPS:                           4499.99
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext ssbd ibrs ibpb stibp vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr arat npt nrip_save umip rdpid
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          32 KiB (1 instance)
L1i cache:                          32 KiB (1 instance)
L2 cache:                           512 KiB (1 instance)
L3 cache:                           16 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0,1
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Vulnerable, no microcode
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.1.0+cu121
[pip3] torchaudio==2.1.0+cu121
[pip3] torchdata==0.7.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.16.0
[pip3] torchvision==0.16.0+cu121
[pip3] triton==2.1.0
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

**Comment by tzielinski-habana** [comment]
@eellison, thanks, do you know if the fix will be included in the 2.2 release?

**Comment by zielin32** [comment]
@eellison, thanks, do you know if the fix will be included in the 2.2 release?

**Comment by eellison** [comment]
@zielin32 sorry i don't think it will be. However you can apply this fix on the user side by unsqueezing the 0-dim indexing tensor `torch.tensor(0).to(device).unsqueeze(0)`
